test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/ViewFeatures/ViewDataDictionaryOfTModelTest.cs

```diff
@@ -223,6 +226,48 @@ public void CopyConstructors_ThrowInvalidOperation_IfModelIncompatible()
             Assert.Equal(expectedMessage, exception.Message);
         }
 
+        [Theory]
+        [InlineData(null)]
+        [InlineData(23)]
+        public void CopyConstructor_DoesNotChangeMetadata_WhenValueCompatibleWithSourceMetadata(int? model)
+        {
+            // Arrange
+            var metadataProvider = new EmptyModelMetadataProvider();
+            var source = new ViewDataDictionary<int?>(metadataProvider)
+            {
+                Model = -48,
+            };
+
+            // Act
+            var viewData = new ViewDataDictionary<int?>(source, model);
+
+            // Assert
+            Assert.NotNull(viewData.ModelExplorer);
+            Assert.NotSame(source.ModelExplorer, viewData.ModelExplorer);
+            Assert.Same(source.ModelMetadata, viewData.ModelMetadata);
+            Assert.Equal(typeof(int?), viewData.ModelMetadata.ModelType);
+            Assert.Equal(viewData.Model, viewData.ModelExplorer.Model);
+            Assert.Equal(model, viewData.Model);
+        }
+
+        [Fact]
+        public void CopyConstructor_UpdatesMetadata_IfDeclaredTypeChangesIncompatibly()
```

Comment:
Should we also have a case here where the `Model` is set and the type changes to something incompatible.